### ISSUE TITLE
Can't use in() on generator outputs.

### ISSUE THREAD

**Issue by abadams** [halide/Halide#6459]
I would like to be able to write the following, where `output` is an `Output<Buffer<>>` member of a generator:
```
output(x, y) += foo;
output.compute_at(output.in(), y)
```

This would need some sort of checking if a pipeline output has a global wrapper in early lowering, and if so, promoting that wrapper to be the actual output.

**Comment by abadams** [comment]
It would also require something clever for generators, because you might want to say output.in().dim(0).set_extent(...) or the like.

**Comment by abadams** [comment]
Or maybe you promote the info from the existing output buffer to apply to the new wrapper Func.

**Comment by abadams** [comment]
Yes, that's why I thought to open the issue as a separate general thing.

**Comment by steven-johnson** [comment]
I will implement this if and only if someone promises to show me where the definitive documentation for how to use `.in()` is, because I still don't really grasp it after all these years. (And if the documentation doesn't exist, ahem.)

**Comment by abadams** [comment]
in() actually has pretty good documentation in the doxygen. It also has a longer tutorial on it:

https://halide-lang.org/tutorials/tutorial_lesson_19_wrapper_funcs.html

**Comment by mcourteaux** [comment]
Oh god, thanks for this tutorial. I'm really getting the hang of `.in()`. And now I understand it, I really see why this is a good issue. I actually really need this now. I can't schedule reductions on output Functions with the zero-init kernel going AND not making an intermediate extra function in the generate().

**Comment by mcourteaux** [comment]
How do we deal with multiple wrapper functions being defined for an output?

```cpp
Output<Buffer<>> output;
output(x, y) += foo;
output.compute_at(output.in(), y); // Global wrapper

Func dependent;
dependent(x, y) = output(x, y);
output.in(dependent).compute_at(y); // Dependent-function-specific wrapper
```

I think it makes only sense to promote the global one to be the output. However, I think the `output.in(dependent)` should take it's data from the global wrapper, right?

**Comment by mcourteaux** [comment]
Actually, this boils down to a more fundamental shortcoming of the scheduling right now I think. As the lowering code does .store_root().compute_root():
https://github.com/halide/Halide/blob/392430d1133ee8cc5f4fe87b4dd06ac9779eb8d8/src/Lower.cpp#L119-L122
We can't schedule two outputs that are dependent without intermediate function, which I think shouldn't be strictly necessary?

```cpp
Output<Buffer<>> output;
output(x, y) = ...;

Output<Buffer<>> dependent;
dependent(x, y) = output(x, y) + 1.0;

output.store_root().compute_at(dependent, y);
dependent.compute_root();
```
IMO, this should be a valid schedule, but isn't possible because of the linked code where Lower forces the use of `.compute_root()`. Maybe there are technical underlying reasons why this is though. I don't know. In other words, what I'm thinking of is that maybe the Lowering should only do `.compute_root()` if no other `.compute_xxx()` directive was used yet on that output Func.

**Comment by abadams** [comment]
The technical reasons are the same as the ones in #6367 and you have to break the dependency and use compute_with in the same way to get that to work.

If compute_with automatically handled things with dependencies, we wouldn't need to do that, but there's some hairy analysis code to write.

In .in() worked on outputs, you could do the transformation like so:

output.in({/* nothing */}).compute_with(dependent, ...)

dependent and output.in() both call the original version of output, so they are eligible to be compute_with each other.

**Comment by abadams** [comment]
Maybe output.in({/* nothing */}) should have a different name, like output.wrap() though.

**Comment by mcourteaux** [comment]
Could you clarify to which schedule you are reacting? I made the mistake to use the same names in both schedules :confused:.

**Comment by abadams** [comment]
This one: 
```
output.store_root().compute_at(dependent, y);
dependent.compute_root();
```

I was trying to imagine what schedule one could write to get two pipeline outputs with a dependency between them into a single loop nest.
